src/Umbraco.Infrastructure/Security/MemberRoleStore.cs

```diff
@@ -0,0 +1,342 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Services;
+
+namespace Umbraco.Cms.Core.Security
+{
+    /// <summary>
+    /// A custom user store that uses Umbraco member data
+    /// </summary>
+    public class MemberRoleStore<TRole> : IRoleStore<TRole> where TRole : IdentityRole
+    {
+        private readonly IMemberGroupService _memberGroupService;
+        private bool _disposed;
+
+        //TODO: Move into custom error describer.
+        //TODO: How revealing can the error messages be?
+        private readonly IdentityError _intParseError = new IdentityError { Code = "IdentityIdParseError", Description = "Cannot parse ID to int" };
+        private readonly IdentityError _memberGroupNotFoundError = new IdentityError { Code = "IdentityMemberGroupNotFound", Description = "Member group not found" };
+        private const string genericIdentityErrorCode = "IdentityErrorUserStore";
+
+        public MemberRoleStore(IMemberGroupService memberGroupService, IdentityErrorDescriber errorDescriber)
+        {
+            _memberGroupService = memberGroupService ?? throw new ArgumentNullException(nameof(memberGroupService));
+            ErrorDescriber = errorDescriber ?? throw new ArgumentNullException(nameof(errorDescriber));
+        }
+
+        /// <summary>
+        /// Gets or sets the <see cref="IdentityErrorDescriber"/> for any error that occurred with the current operation.
+        /// </summary>
+        public IdentityErrorDescriber ErrorDescriber { get; set; }
+
+        /// <inheritdoc />
+        public Task<IdentityResult> CreateAsync(TRole role, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                ThrowIfDisposed();
+
+                if (role == null)
+                {
+                    throw new ArgumentNullException(nameof(role));
```

Comment:
**nitpick:** consider letting these exceptions throw. I wouldn't return an `IdentityResult` for things like this. If the store manages to be called with a null role, then I'd say that's exceptional behavior. I'm not sure what the convention is though

Reply:
Good point @scottbrady91 - @bergmania  @Shazwazza  what is our preference here? Let the exception throw, or catch and return a failed identity result like:

`return Task.FromResult(IdentityResult.Failed(new IdentityError { Code = genericIdentityErrorCode, Description = ex.Message }));`

Reply:
IMO an `ArgumentNullException` should throw..

Reply:
Thanks, have removed the try catch